physicalplan/scan/matrix_selector.go

```diff
@@ -127,19 +127,44 @@ func (o *matrixSelector) Next(ctx context.Context) ([]model.StepVector, error) {
 		)
 
 		for currStep := 0; currStep < numSteps && seriesTs <= o.maxt; currStep++ {
-			if len(vectors) <= currStep {
-				vectors = append(vectors, o.vectorPool.GetStepVector(seriesTs))
-			}
 			maxt := seriesTs
 			mint := maxt - o.selectRange
-
 			rangePoints := selectPoints(series.samples, mint, maxt, o.scanners[i].previousPoints)
-			result := o.call(series.labels, rangePoints, time.UnixMilli(seriesTs), time.Duration(o.selectRange)*time.Millisecond)
-			if result.Point != InvalidSample.Point {
-				vectors[currStep].T = result.T
-				vectors[currStep].Samples = append(vectors[currStep].Samples, result.V)
-				vectors[currStep].SampleIDs = append(vectors[currStep].SampleIDs, series.signature)
+
+			if o.in != nil {
+				if len(vectors) <= currStep {
+					vectors = append(vectors, o.vectorPool.GetStepVector(seriesTs))
+				}
+
+				// TODO(saswatamcode): Handle multi-arg functions for matrixSelectors via injectable.
+				result := o.in.Call(function.FunctionArgs{
+					Labels:      series.labels,
+					Points:      rangePoints,
+					StepTime:    time.UnixMilli(seriesTs),
+					SelectRange: time.Duration(o.selectRange) * time.Millisecond,
+				})
+
+				if result.Point != function.InvalidSample.Point {
+					vectors[currStep].T = result.T
+					vectors[currStep].Samples = append(vectors[currStep].Samples, result.V)
+					vectors[currStep].SampleIDs = append(vectors[currStep].SampleIDs, series.signature)
+				}
+			} else {
```

Comment:
Let's leave this out for now and revert this change. We can fall back to the old engine in this case and maybe open an issue to tackle it in the future.

Reply:
Yup, issue is already open #39. I'll add a TODO for now. This is also why I added the injectable type.